internal/agent/configcontroller/enrollment.go

```diff
@@ -0,0 +1,152 @@
+package configcontroller
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/flightctl/flightctl/api/v1alpha1"
+	"github.com/flightctl/flightctl/client"
+	"github.com/skip2/go-qrcode"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/cert"
+	"k8s.io/klog/v2"
+)
+
+func (c *ConfigController) ensureDeviceEnrollment(ctx context.Context, device *v1alpha1.Device) error {
+	if c.isDeviceEnrolled() {
+		return nil
+	}
+
+	if err := c.deviceEnrollmentRequest(ctx, device); err != nil {
+		return fmt.Errorf("failed to send enrollment request: %w", err)
+	}
+
+	klog.Infof("%swaiting for enrollment to be approved", c.logPrefix)
+	err := wait.ExponentialBackoff(c.enrollmentVerifyBackoff, func() (bool, error) {
+		return c.verifyDeviceEnrollment(ctx, device)
+	})
+	if err != nil {
+		return fmt.Errorf("failed to verify device enrollment: %w", err)
+	}
+	if err := c.writeDeviceEnrollmentBanner(ctx, device); err != nil {
```

Comment:
This needs to happen before waiting for approval, as it's pre-requisite for the approval process.